### DISCUSSION TITLE
Show progress status for oras commands

### DISCUSSION THREAD

**Discussion by qweeah** [oras-project/oras#1042] [General]
# Problem
As is pointed out in https://github.com/oras-project/oras/issues/940 and https://github.com/oras-project/oras/issues/893, oras commands might halt when processing data transfer tasks without any user feedback.
# Solution
If ORAS user is using a terminal to run the command, we can compose an UI in the terminal and display progress status of the execution.

## Goals
- Support progress status output for the following commands 
  - oras push
  - oras attach
  - oras pull
  - oras cp
  - oras blob fetch
  - oras blob push
  - ~~oras manifest fetch~~
  - oras manifest push
- Introduce only necessary and stable (>=1.0.0) dependencies
- Display key information to reflect the progress (TBD: what to display & how to display)
  - Total size
  - Transferred size
  - Percentage
  - Time taken
  - Task status: running or done
  - Progress bar
  - Others: please comment

## Non-goals
- Improve output readability: the status format should be similar to [v1.0.0](https://github.com/oras-project/oras/releases/tag/v1.0.0)
- Provide interactable UI: the output should be read-only. User cannot interact via clicking mouse or hovering over an area
- Support terminal with width less than 80 characters (width of [vt100](https://en.wikipedia.org/wiki/VT100))
- Support non-ANSI characters, e.g. emoji

# PoC
I had an implementation in my own fork based on below MVC design:
```mermaid
flowchart LR
subgraph "oras CLI"
    subgraph root
        direction LR
        C[console]
        A[ANSI control]
    end
    
    subgraph progress
        S["Status(model)"]--rendered by-->M["Manager(view)"]
        M-->A
        M-->C
    end
    
    subgraph "tracker(controller)"
        T[Target]--manipulates-->S
        R[Reader]--manipulates-->S
    end
    
    subgraph commands
        push-->T
        attach-->T
        pull-->T
        cp-->T
        bf[blob fetch]-->R
        mf[manifest fetch]-->R
        bp[blob push]-->R
        mp[manifest push]-->R
    end
end
```

I cut three releases with different UI:
- spinner: https://github.com/qweeah/oras/releases/tag/v1.2.0-spinner
[![asciicast](https://asciinema.org/a/602698.svg)](https://asciinema.org/a/602698)

- progress bar: https://github.com/qweeah/oras/releases/tag/v1.2.0-bar
[![asciicast](https://asciinema.org/a/cA9PyAzBf8zQeoDTaCnaG1ASf.svg)](https://asciinema.org/a/cA9PyAzBf8zQeoDTaCnaG1ASf)

- two-line output with progress bar: https://github.com/qweeah/oras/releases/tag/v1.2.0-2line-bar
  (This implementation breaks current output style and split output of one blob into 2 lines)
[![asciicast](https://asciinema.org/a/603318.svg)](https://asciinema.org/a/603318)

To try out more commands, you may installed via below command
```shell
# example for linux/amd64
VERSION="1.2.0-spinner" # 1.2.0-2line or 1.2.0-bar
OSTYPE="linux"
ARCH="amd64"
curl -LO "https://github.com/qweeah/oras/releases/download/v${VERSION}/oras_${VERSION}_${OSTYPE}_${ARCH}.tar.gz"
mkdir -p oras-install/
tar -zxf oras_${VERSION}_*.tar.gz -C oras-install/
sudo mv oras-install/oras /usr/local/bin/oras-spinner
rm -rf oras_${VERSION}_*.tar.gz oras-install/
```

**Comment by TerryHowe** [comment]
If we had to choose one, I think I'd slightly prefer spinner

**Reply by qweeah** [reply]
Can you help elaborate more on why the barred output looks less preferrable?

Also, what would be a good UI in your mind?

**Reply by sajayantony** [reply]
I think the spinner is more subtle. If the bar were less dense like docker progress bar then maybe it might appeal.

**Reply by qweeah** [reply]
Docker progress bar requires more horizontal space than oras CLI can offer. If we can create a new line to display the progress bar below the status, then it's doable. I can add another implementation to demo that.

**Reply by TerryHowe** [reply]
I like the way you minimized the space with the bar and it is more informative, but I like the spinner because you know you have progress without being overwhelmed.

**Reply by qweeah** [reply]
It's hard to choose a right background color for the progress bar because I don't find any suitable package to deal with terminal theming, so I just use the foreground color, which doesn't look good.

**Comment by qweeah** [comment]
Added a new implementation which separated the output of progress bar into a newline @sajayantony @TerryHowe any feedback?
[![asciicast](https://asciinema.org/a/Rxcrsjh4OYeLZT77tll5hHA5F.svg)](https://asciinema.org/a/Rxcrsjh4OYeLZT77tll5hHA5F)

**Reply by qweeah** [reply]
Added another implementation to show NPM-like progress bar
[![asciicast](https://asciinema.org/a/qKQ9hEhMV5tfT8nVfyjdR9idZ.svg)](https://asciinema.org/a/qKQ9hEhMV5tfT8nVfyjdR9idZ)

**Comment by TerryHowe** [comment]
Even though it uses more vertical space, I like that implementation better than the tighter one. I like the npm style one less, but I just haven't done a lot of javascripting!

**Comment by FeynmanZhou** [comment]
It might be helpful to display the time usage for uploading/downloading each blob in the progress bar.

**Reply by qweeah** [reply]
Added, please check the demo in below:
[![asciicast](https://asciinema.org/a/603318.svg)](https://asciinema.org/a/603318)

**Comment by shizhMSFT** [comment]
The two-line one LGTM. It would be better if the cells for progress in size, %, and time are aligned.

**Reply by qweeah** [reply]
Added a new implementation to align the output with instant speed indication.
[![asciicast](https://asciinema.org/a/lhoEt0gyVfiRBqfKETqKrYnRY.svg)](https://asciinema.org/a/lhoEt0gyVfiRBqfKETqKrYnRY)
